src/bvals/comms/coalesced_buffers.cpp

```diff
@@ -0,0 +1,482 @@
+
+  int nbnd_id{0};
+  int comb_size{0};
+  for (auto uid : var_set) {
+    // Skip this variable if it is not communicated in this BoundaryType
+    if (coalesced_info_buf.count(uid) == 0) continue;
+    nbnd_id += coalesced_info_buf.at(uid).size();
+    for (auto &[bnd_id, pvbbuf] : coalesced_info_buf.at(uid)) {
+      auto buf_state = pvbbuf->GetState();
+      if ((buf_state == BufferState::sending) || (buf_state == BufferState::received))
+        comb_size += bnd_id.size();
+    }
+  }
+  if (pcomb_size != nullptr) *pcomb_size = comb_size;
+
+  bool updated = false;
+  if (comb_size > coalesced_comm_buffer.buffer().size()) {
+    PARTHENON_REQUIRE(
+        sender, "Something bad is going on if we are doing this on a receiving buffer.");
+    coalesced_comm_buffer.ConstructBuffer("combined send buffer", 2 * comb_size);
+    updated = true;
+  }
+
+  if (nbnd_id != bnd_ids_device.size()) {
+    bnd_ids_device = ParArray1D<BndId>("bnd_id", nbnd_id);
+    bnd_ids_host = Kokkos::create_mirror_view(bnd_ids_device);
+    updated = true;
+  }
+
+  int idx{0};
+  int c_buf_idx{0}; // Index at which v-b buffer starts in combined buffer
+  for (auto uid : var_set) {
+    // Skip this variable if it is not communicated in this BoundaryType
+    if (coalesced_info_buf.count(uid) == 0) continue;
+    for (auto &[bnd_id, pvbbuf] : coalesced_info_buf.at(uid)) {
+      auto &bid_h = bnd_ids_host[idx];
+      auto buf_state = pvbbuf->GetState();
+      PARTHENON_REQUIRE(buf_state != BufferState::stale,
+                        "Trying to work with a stale buffer.");
+
+      const bool alloc =
+          (buf_state == BufferState::sending) || (buf_state == BufferState::received);
+
+      // Test if this boundary has changed
+      if (!bid_h.SameBVChannel(bnd_id) || (bid_h.buf_allocated != alloc) ||
+          (bid_h.start_idx() != c_buf_idx) ||
+          !UsingSameResource(bid_h.buf, pvbbuf->buffer()) ||
+          bid_h.coalesced_buf.data() != coalesced_comm_buffer.buffer().data()) {
+        updated = true;
+        bid_h = bnd_id;
+        bid_h.buf_allocated = alloc;
+        bid_h.start_idx() = c_buf_idx;
+        bid_h.coalesced_buf = coalesced_comm_buffer.buffer();
+        if (bid_h.buf_allocated) bid_h.buf = pvbbuf->buffer();
+      }
+      if (bid_h.buf_allocated) c_buf_idx += bid_h.size();
+      idx++;
+    }
+  }
+  if (updated) Kokkos::deep_copy(bnd_ids_device, bnd_ids_host);
+  return bnd_ids_device;
+}
+
+//----------------------------------------------------------------------------------------
+void CoalescedBuffer::PackAndSend(const std::set<Uid_t> &vars) {
+  PARTHENON_REQUIRE(coalesced_comm_buffer.IsAvailableForWrite(),
+                    "Trying to write to a buffer that is in use.");
+  int comb_size;
+  auto &bids = GetBndIdsOnDevice(vars, &comb_size);
+  Kokkos::parallel_for(
+      PARTHENON_AUTO_LABEL,
+      Kokkos::TeamPolicy<>(parthenon::DevExecSpace(), bids.size(), Kokkos::AUTO),
```

Comment:
`bids` is potentially a low number (lower than the number of compute unites per device), isn't it?
Just so that we keep this in mind wrt the kernels' efficiency.

Reply:
yes, this could definitely be somewhere we don't produce enough teams.